Version 1.0

Aleks1795 opened the pull request:
- Retrofit implemented
- NO loading bar
- NO cache

Aleks1795 committed:
Empty lines deleted

Aleks1795 committed:
Version 1.0
- Retrofit implemented
- NO loading bar
- NO cache

RuslanMingaliev commented:
PR name should reflect the changes. For example, "[FR-1] Add cats facts"

RuslanMingaliev commented on app/src/main/java/com/ruslanchikbananchik/rmma/common/Common.kt:
```diff
@@ -0,0 +1,12 @@
+object Common {
+    private val BASE_URL = "https://meowfacts.herokuapp.com/"
+    val retrofitService: RetrofitServices
+        get() = RetrofitClient.getClient(BASE_URL).create(RetrofitServices::class.java)
```
Why did you use `get()` here? With this `get()` a new instance of `RetrofitService` will be created on each call. You don't need that.

You tried to hide implementation details of `RetrofitService` creation here. This approach is called "Factory method" pattern. You can rename this class from `Common` to `RetrofitServiceFactory` and add `fun create(): RetrofitService` function. It would be clearer for other devs who will read this code.

RuslanMingaliev commented on app/src/main/java/com/ruslanchikbananchik/rmma/interface/RetrofitServices.kt:
```diff
@@ -0,0 +1,11 @@
+
+interface RetrofitServices {
+    @GET("?count=3")
+
```
Codestyle violation 

https://kotlinlang.org/docs/coding-conventions.html#annotations
remove this empty line

RuslanMingaliev commented on app/src/main/java/com/ruslanchikbananchik/rmma/model/Fact.kt:
```diff
@@ -0,0 +1,5 @@
+package com.ruslanchikbananchik.rmma.model
+
+data class Fact(
+    var fact: String? = null
```
Why it's nullable?

RuslanMingaliev commented on app/src/main/java/com/ruslanchikbananchik/rmma/retrofit/RetrofitClient.kt:
```diff
@@ -0,0 +1,19 @@
+    private var retrofit: Retrofit? = null
+
+    fun getClient(baseUrl: String): Retrofit {
+        if (retrofit == null) {
```
Not thread-safe implementation. See https://en.wikipedia.org/wiki/Singleton_pattern
In Kotlin you can use `by lazy { ... }` delegate. See https://kotlinlang.org/docs/delegated-properties.html#lazy-properties. It creates an instance on the first call and is thread-safe by default.

RuslanMingaliev commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -6,14 +6,13 @@
-        android:layout_width="wrap_content"
+    <androidx.recyclerview.widget.RecyclerView
+        android:id="@+id/factsRecyclerView"
+        android:layout_width="match_parent"
```
As I remember, you have to use `0dp` except of `match_parent` if you set constraints (`app:layout_constraintStart_toStartOf`, `layout_constraintEnd_toEndOf`). The same with the height. Android Studio should highlight this value and suggest to change it.

RuslanMingaliev replied:
FrameLayout is good here too.

RuslanMingaliev commented on app/src/main/res/layout/item_layout.xml:
```diff
@@ -0,0 +1,30 @@
+    app:cardElevation="8dp"
+    android:layout_margin="8dp">
+
+    <androidx.constraintlayout.widget.ConstraintLayout
```
`ConstraintLayout` is too complex for only one TextView. Use `FrameLayout` here.

RuslanMingaliev commented on app/src/main/res/values/strings.xml:
```diff
@@ -1,4 +1,6 @@
-    <string name="hello_android">Hello Android! Version: %s</string>
+    <string name="catFact_1">Cat Fact #1: %s</string>
+    <string name="catFact_2">Cat Fact #2: %s</string>
+    <string name="catFact_3">Cat Fact #3: %s</string>
```
It's not a good idea to add a new string for every fact. What about 500 facts? 
You can use `Cat Fact: #%n: %s` and use it like `Context#getString(1, "Fact 1")`

RuslanMingaliev commented on app/build.gradle:
```diff
@@ -1,6 +1,8 @@
     id 'com.android.application'
     id 'org.jetbrains.kotlin.android'
+    id 'kotlin-android'
+    id 'kotlin-android-extensions'
```
It's deprecated

RuslanMingaliev commented on app/src/main/java/com/ruslanchikbananchik/rmma/MainActivity.kt:
```diff
@@ -3,17 +3,52 @@ package com.ruslanchikbananchik.rmma
+    }
+
+    private fun getCatFacts() {
+        fService.getFactsList().enqueue(object : Callback<MutableList<Fact>> {
```
Try to rotate the screen while loading

RuslanMingaliev commented on app/src/main/java/com/ruslanchikbananchik/rmma/interface/RetrofitServices.kt:
```diff
@@ -0,0 +1,11 @@
+interface RetrofitServices {
+    @GET("?count=3")
+
+    fun getFactsList(): Call<MutableList<Fact>>
```
Don't need to use `MutableList` here, `List` is enough

RuslanMingaliev commented on app/src/main/java/com/ruslanchikbananchik/rmma/MainActivity.kt:
```diff
@@ -3,17 +3,52 @@ package com.ruslanchikbananchik.rmma
 import android.os.Bundle
+import androidx.recyclerview.widget.LinearLayoutManager
+import androidx.recyclerview.widget.RecyclerView.LayoutManager
+import com.ruslanchikbananchik.rmma.`interface`.RetrofitServices
```
it's not a good idea to use keywords

Aleks1795 committed:
Fixed:
- App works correctly (Cat facts are shown on the app launch as it was designed "Cat Fact #x: Here's the fact")
- "No connection" loading spinner is added

RuslanMingaliev commented:
Any updates on this?
